### ISSUE TITLE
Count-trailing-ones emulation emit is suboptimal

### ISSUE THREAD

**Issue by Validark** [llvm/llvm-project#82487]
For my RISC-V Sifive-u74, LLVM emits David Seal's algorithm for count-trailing-zeroes emulation, since it lacks nice bit manipulation instructions.

Unfortunately, when I write a count-trailing-ones function, it misses the obvious optimization. In Zig: [Godbolt link](https://zig.godbolt.org/z/Tac6YbT1q)

```zig
export fn ct1(x: u64) u8 {
    return @ctz(~x);
}
```

```asm
.LCPI0_0:
        .quad   151050438420815295
.LCPI0_1:
        .ascii  "\000\001\002\007\003\r\b\023\004\031\016\034\t\"\024(\005\021\032&\017.\0350\n\037#6\0252)9?\006\f\022\030\033!'\020%-/\036518>\013\027 $,47=\026+3<*;:"
ctz:
        li      a1, -1
        beq     a0, a1, .LBB0_2
        lui     a1, %hi(.LCPI0_0)
        not     a0, a0
        neg     a2, a0
        and     a0, a0, a2
        ld      a1, %lo(.LCPI0_0)(a1)
        mul     a0, a0, a1
        lui     a1, %hi(.LCPI0_1)
        srli    a0, a0, 58
        addi    a1, a1, %lo(.LCPI0_1)
        add     a0, a0, a1
        lbu     a0, 0(a0)
        ret
.LBB0_2:
        li      a0, 64
        ret
```

The same issue occurs in C as well: [Godbolt link](https://godbolt.org/z/oxKPWKnTs)

```c
uint64_t ctz(uint64_t x) {
    if (x == 0) return 64;
    return __builtin_ctzll(x);
}

uint64_t ct1(uint64_t x) {
    return ctz(~x);
}
```

```asm
.LCPI1_0:
        .quad   151050438420815295              # 0x218a392cd3d5dbf
.LCPI1_1:
        .ascii  "\000\001\002\007\003\r\b\023\004\031\016\034\t\"\024(\005\021\032&\017.\0350\n\037#6\0252)9?\006\f\022\030\033!'\020%-/\036518>\013\027 $,47=\026+3<*;:"
ct1(unsigned long):                                # @ct1(unsigned long)
        li      a1, -1
        beq     a0, a1, .LBB1_2
.Lpcrel_hi2:
        auipc   a1, %pcrel_hi(.LCPI1_0)
        not     a0, a0
        neg     a2, a0
        and     a0, a0, a2
        ld      a1, %pcrel_lo(.Lpcrel_hi2)(a1)
        mul     a0, a0, a1
.Lpcrel_hi3:
        auipc   a1, %pcrel_hi(.LCPI1_1)
        srli    a0, a0, 58
        addi    a1, a1, %pcrel_lo(.Lpcrel_hi3)
        add     a0, a0, a1
        lbu     a0, 0(a0)
        ret
.LBB1_2:
        li      a0, 64
        ret
```

David Seal's algorithm starts by checking our input, `x`, against 0. If `x` is 0, we return 64 (for 64-bit count-trailing zero). Then, we compute `y = x & -x`, where `y` contains just the lowest `1` bit of `x`. Thus, `@ctz(y) ≡ @ctz(x)`, but crucially, there are only 64 possible values of `y` (64-bit bitstrings with a popcount of 1). The rest of the algorithm is a perfect hash into a lookup table.

If we trim down the algorithm, LLVM finds the optimization. [Godbolt link](https://zig.godbolt.org/z/v61Efe4Gh)

```zig
export fn ctzStart(x: u64) u64 {
    if (x == 0) return 64;
    return x & (~x +% 1);
}

export fn ct1Start(x: u64) u64 {
    return ctzStart(~x);
}
```

```asm
ctzStart:
        neg     a2, a0
        li      a1, 64
        beqz    a0, .LBB0_2
        and     a1, a2, a0
.LBB0_2:
        mv      a0, a1
        ret

ct1Start:
        addi    a1, a0, 1
        not     a2, a0
        li      a0, 64
        beqz    a1, .LBB1_2
        and     a0, a1, a2
.LBB1_2:
        ret
```

The optimization is when we insert `~x` in for `x` in `x & -x`, getting `~x & -~x`. It happens that `-~x` is the same as `x + 1`. It also happens that checking `~x == 0` can be performed by checking `x + 1 == 0` or, ya know, `~x == 0`. In the count-trailing-ones function, LLVM instead loads a `-1` constant and compares against that, which can't be used for anything else. It also takes the bitwise inverse of `x` (`n = ~x`), and feeds that into `n & -n`. This is worse than simply doing `~x & (x + 1)` because `x + 1` and `~x` may be computed in parallel, whereas with `~x & -~x`, the hardware probably isn't going to fold `-~x` into an increment.

This is the emit I get with the optimization written out manually:

```asm
.LCPI0_0:
        .quad   151050438420815295
ct1:
        li      a1, 64
        addi    a2, a0, 1
        beqz    a2, .LBB0_2
        lui     a1, %hi(.LCPI0_0)
        not     a0, a0
        and     a0, a0, a2
        ld      a1, %lo(.LCPI0_0)(a1)
        mul     a0, a0, a1
        lui     a1, %hi(.L__unnamed_1)
        srli    a0, a0, 58
        addi    a1, a1, %lo(.L__unnamed_1)
        add     a0, a0, a1
        lbu     a1, 0(a0)
.LBB0_2:
        mv      a0, a1
        ret

.L__unnamed_1:
        .ascii  "\000\001\002\007\003\r\b\023\004\031\016\034\t\"\024(\005\021\032&\017.\0350\n\037#6\0252)9?\006\f\022\030\033!'\020%-/\036518>\013\027 $,47=\026+3<*;:"
```

**Comment by Validark** [comment]
Alternatively, could you turn count trailing zeroes into David Seal's algorithm earlier in the pipeline? I guess that could disable certain optimizations you might be able to do with the knowledge that David Seal finds the trailing-zero count, but it would enable more things to be folded into the emulation of this operation.

But maybe then that wouldn't perfectly square with https://github.com/llvm/llvm-project/issues/80100, where I hope we could use a faster emulation routine on cases with more restrictions on the input.

Tricky to have it both ways, I suppose, haha.

**Comment by Validark** [comment]
A similar issue happens on sparc64 niagara4, which implements `ctz(x)` with `popcount(~x & x - 1)`. Count-trailing-ones can be `x & (-2 - x)` or `x & ~(x + 1)` or `x & (~x - 1)`.
